File: Examples/Plugins/HioImage/hioPpm_Swift/Sources/hioPpm/ppmImage.swift
//===----------------------------------------------------------------------===//
// This source file is part of github.com/apple/SwiftUsd
//
// Copyright © 2025 Apple Inc. and the SwiftUsd project authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//  https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0
//===----------------------------------------------------------------------===//

import Foundation
import OpenUSD

extension PpmImage {
    /// Error type for errors while parsing a PPM image
    fileprivate struct ParseError: Error {
        var message: String
        
        init(_ message: String) {
            self.message = message
        }
        
        var description: String { message }
    }
}

/// Class for reading and writing PPM images
class PpmImage {
    private(set) var width: UInt16
    private(set) var height: UInt16
    private(set) var componentMaximum: UInt16
    private var storage: [UInt16]
        
    init?(forReading filename: std.string) {
        do throws(ParseError) {
            
            let asset = Overlay.ArGetResolver().OpenAsset(pxr.ArResolvedPath(filename))
            guard Bool(asset) else {
                throw ParseError("PpmImage createForReading failed during ArResolver::OpenAsset '\(filename)'")
            }
            
            var width: UInt16?
            var height: UInt16?
            var componentMaximum: UInt16?
            var storage: [UInt16] = []
            
            // Parse the file contents
            try withExtendedLifetime((asset, asset.GetBuffer())) { (asset, bufSharedPtr) throws(ParseError) in
                let buf = bufSharedPtr.__getUnsafe()!
                
                var index = 0
                while index < asset.GetSize() {
                    // Check for the magic number at the start of the file
                    if index == 0 {
                        guard buf[index] == UInt8(ascii: "P") else {
                            throw ParseError("Invalid PPM magic number")
                        }
                        index += 1
                        continue
                    }
                    if index == 1 {
                        guard buf[index] == UInt8(ascii: "3") else {
                            throw ParseError("Invalid PPM magic number")
                        }
                        index += 1
                        continue
                    }
                    
                    
                    // Line comments are ignored
                    if buf[index] == UInt8(ascii: "#") && index > 0 && buf[index - 1] == UInt8(ascii: "\n") {
                        // Advance through the end of the line
                        while index < asset.GetSize() && buf[index] != UInt8(ascii: "\n") {
                            index += 1
                        }
                        continue
                    }
                    
                    // Whitespace delimits values
                    if isspace(Int32(buf[index])) != 0 {
                        // Advance through the last contiguous whitespace
                        while index < asset.GetSize() && isspace(Int32(buf[index])) != 0 {
                            index += 1
                        }
                        continue
                    }
                    
                    // Handle numbers in the file
                    if isdigit(Int32(buf[index])) != 0 {
                        var digitBuffer = ""
                        while index < asset.GetSize() && isdigit(Int32(buf[index])) != 0 {
                            digitBuffer += String(UnicodeScalar(UInt8(buf[index])))
                            index += 1
                        }
                        
                        guard let parsedInt = UInt16(digitBuffer) else {
                            throw ParseError("Invalid PPM number (failed to convert)")
                        }
                        
                        if width == nil {
                            guard parsedInt > 0 else {
                                throw ParseError("Invalid PPM width")
                            }
                            width = parsedInt
                            continue
                        }
                        
                        if height == nil {
                            guard parsedInt > 0 else {
                                throw ParseError("Invalid PPM height")
                            }
                            height = parsedInt
                            continue
                        }
                        
                        if componentMaximum == nil {
                            guard parsedInt > 0 else {
                                throw ParseError("Invalid PPM component maximum")
                            }
                            componentMaximum = parsedInt
                            continue
                        }
                        
                        guard parsedInt <= componentMaximum! else {
                            throw ParseError("Invalid PPM, raster component greater than component maximum")
                        }
                        
                        storage.append(parsedInt)
                        guard storage.count <= 3 * width! * height! else {
                            throw ParseError("Invalid PPM, raster larger than expected")
                        }
                        
                        continue
                    } // `if isdigit(Int32(buf[index])) != 0`
                    
                    // We've handled comments, whitespace, and numbers, so
                    // anything else must be an invalid character in a single-image PPM file
                    throw ParseError("Invalid PPM file contents")
                } // outermost `while index < asset.GetSize()`
                
            } // `withExtendedLifetime`
            
            guard let width, let height, let componentMaximum else {
                throw ParseError("Invalid PPM file header")
            }
            
            guard storage.count == 3 * width * height else {
                throw ParseError("Invalid PPM, raster size is unexpected")
            }
            
            self.width = width
            self.height = height
            self.componentMaximum = componentMaximum
            self.storage = storage
            
        } catch let error as ParseError {
            TF_RUNTIME_ERROR(std.string(error.message))
            return nil
        }
    }
    
    init(forWritingWithWidth width: UInt16, height: UInt16, componentMaximum: UInt16) {
        self.width = width
        self.height = height
        self.componentMaximum = componentMaximum
        self.storage = .init(repeating: 0, count: Int(width) * Int(height) * 3)
    }
    
    var isTwoBytesPerComponent: Bool { componentMaximum >= 256 }
    
    func getPixel(x: UInt16, y: UInt16, channel: Int) -> UInt16 {
        guard let index = _toIndex(x: x, y: y, channel: channel) else { return 0 }
        return storage[index]
    }
    
    func setPixel(x: UInt16, y: UInt16, channel: Int, value: UInt16) {
        guard let index = _toIndex(x: x, y: y, channel: channel) else { return }
        guard value <= componentMaximum else {
            TF_CODING_ERROR("Cannot PpmImage.setPixel to a value greater than the component maximum")
            return
        }
        storage[index] = value
    }
    
    func write(filename: std.string) {
        var asset = Overlay.ArGetResolver().OpenAssetForWrite(pxr.ArResolvedPath(filename), .Replace)
        guard Bool(asset) else {
            TF_RUNTIME_ERROR(std.string("PpmImage write failed during ArResolver::OpenAssetForWrite '\(filename)'"))
            return
        }
        
        var lines = [String]()
        lines.append("P3")
        lines.append("\(width) \(height)")
        lines.append("\(componentMaximum)")
        for i in stride(from: 0, to: storage.count, by: 3) {
            lines.append("\(storage[i]) \(storage[i + 1]) \(storage[i + 2])")
        }
        let s = lines.joined(separator: "\n") + "\n"
        

        asset.Write(s, s.count, 0)
        #TF_VERIFY(asset.Close())
    }
    
    private func _toIndex(x: UInt16, y: UInt16, channel: Int) -> Int? {
        if x >= width || y >= height || channel >= 3 || channel < 0 {
            TF_CODING_ERROR("Invalid coordinates or channel for PpmImage")
            return nil
        }
        
        return Int(y) * 3 * Int(width) + 3 * Int(x) + channel
    }
}

File: Examples/Plugins/HioImage/hioPpm_Cxx/Sources/hioPpm/PpmImage.cpp
//===----------------------------------------------------------------------===//
// This source file is part of github.com/apple/SwiftUsd
//
// Copyright © 2025 Apple Inc. and the SwiftUsd project authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//  https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0
//===----------------------------------------------------------------------===//


#include "PpmImage.h"
#include "Diagnostics.h"
#include "pxr/base/tf/diagnostic.h"
#include "pxr/usd/ar/resolver.h"
#include "pxr/usd/ar/writableAsset.h"
#include "pxr/usd/ar/asset.h"

#include <sstream>

/* static */
std::unique_ptr<PpmImage> PpmImage::createForReading(const std::string &filename) {
    std::shared_ptr<pxr::ArAsset> asset = pxr::ArGetResolver().OpenAsset(pxr::ArResolvedPath(filename));
    if (!asset) {
        MY_RUNTIME_ERROR("PpmImage createForReading failed during ArResolver::OpenAsset '" + filename + "'");
        return {};
    }
    
    std::optional<uint16_t> _width = std::nullopt;
    std::optional<uint16_t> _height = std::nullopt;
    std::optional<uint16_t> _componentMaximum = std::nullopt;
    std::vector<uint16_t> _storage;
    
    // Parse the file contents
    std::shared_ptr<const char> bufSharedPtr = asset->GetBuffer();
    const char* buf = bufSharedPtr.get();
    size_t index = 0;
    while (index < asset->GetSize()) {
        // Check for the magic number at the start of the file
        if (index == 0) {
            if (buf[index] != 'P') {
                MY_RUNTIME_ERROR("Invalid PPM magic number");
                return {};
            } else {
                index += 1;
                continue;
            }
        }
        
        if (index == 1) {
            if (buf[index] != '3') {
                MY_RUNTIME_ERROR("Invalid PPM magic number");
                return {};
            } else {
                index += 1;
                continue;
            }
        }

        // Line comments are ignored
        if (buf[index] == '#' && index > 0 && buf[index - 1] == '\n') {
            while (index < asset->GetSize() && buf[index] != '\n') {
                index += 1;
            }
            continue;
        }
        
        // Whitespace delimits values
        if (isspace(buf[index])) {
            while (index < asset->GetSize() && isspace(buf[index])) {
                index += 1;
            }
            continue;
        }
        
        // Handle numbers in the file
        if (isdigit(buf[index])) {
            std::string digitBuffer;
            while (index < asset->GetSize() && isdigit(buf[index])) {
                digitBuffer += buf[index];
                index += 1;
            }
            
            uint16_t parsedInt;
            try {
                std::int32_t x = std::stoi(digitBuffer);
                if (x < 0 || x > static_cast<int32_t>(std::numeric_limits<uint16_t>::max())) {
                    MY_RUNTIME_ERROR("Invalid PPM number (out of range)");
                    return {};
                }
                parsedInt = static_cast<uint16_t>(x);
            } catch (...) {
                MY_RUNTIME_ERROR("Invalid PPM number (failed to convert)");
                return {};
            }
            
            if (!_width.has_value()) {
                _width = parsedInt;
                if (*_width == 0) {
                    MY_RUNTIME_ERROR("Invalid PPM width");
                    return {};
                }
                continue;
            }
            
            if (!_height.has_value()) {
                _height = parsedInt;
                if (*_height == 0) {
                    MY_RUNTIME_ERROR("Invalid PPM height");
                    return {};
                }
                continue;
            }
            
            if (!_componentMaximum.has_value()) {
                _componentMaximum = parsedInt;
                if (*_componentMaximum == 0) {
                    MY_RUNTIME_ERROR("Invalid PPM component maximum");
                    return {};
                }
                continue;
            }

            if (parsedInt > *_componentMaximum) {
                MY_RUNTIME_ERROR("Invalid PPM, raster component greater than component maximum");
                return {};
            }
            _storage.push_back(parsedInt);
            if (_storage.size() > 3 * (*_width) * (*_height)) {
                MY_RUNTIME_ERROR("Invalid PPM, raster larger than expected");
                return {};
            }
            
            continue;
        }
        
        // We've handled comments, whitespace, and numbers, so
        // anything else must be an invalid character in a single-image PPM file
        MY_RUNTIME_ERROR("Invalid PPM file contents");
        return {};
    }
    
    if (!_width.has_value() || !_height.has_value() || !_componentMaximum.has_value()) {
        MY_RUNTIME_ERROR("Invalid PPM file header");
        return {};
    }
    if (_storage.size() != 3 * (*_width) * (*_height)) {
        MY_RUNTIME_ERROR("Invalid PPM, raster size is unexpected");
        return {};
    }
    
    std::unique_ptr<PpmImage> result = std::unique_ptr<PpmImage>(new PpmImage());
    result->_width = *_width;
    result->_height = *_height;
    result->_componentMaximum = *_componentMaximum;
    result->_storage = std::move(_storage);
    
    return result;
}

/* static */
std::unique_ptr<PpmImage> PpmImage::createForWriting(uint16_t width, uint16_t height, uint16_t componentMaximum) {
    std::unique_ptr<PpmImage> _result = std::unique_ptr<PpmImage>(new PpmImage());
    _result->_width = width;
    _result->_height = height;
    _result->_componentMaximum = componentMaximum;
    _result->_storage.resize(width * height * 3);
    return _result;
}

bool PpmImage::isTwoBytesPerComponent() const {
    return _componentMaximum >= 256;
}

uint16_t PpmImage::getPixel(uint16_t x, uint16_t y, int channel) const {
    size_t index = _toIndex(x, y, channel);
    if (index == -1) { return 0; }
    return _storage[index];
}

void PpmImage::setPixel(uint16_t x, uint16_t y, int channel, uint16_t value) {
    size_t index = _toIndex(x, y, channel);
    if (index == -1) { return; }
    if (value > _componentMaximum) {
        MY_CODING_ERROR("Cannot PpmImage::setPixel to a value greater than the component maximum");
        return;
    }
    _storage[index] = value;
}

void PpmImage::write(const std::string &filename) {
    std::shared_ptr<pxr::ArWritableAsset> asset = pxr::ArGetResolver().OpenAssetForWrite(pxr::ArResolvedPath(filename), pxr::ArResolver::WriteMode::Replace);
    if (!asset) {
        MY_RUNTIME_ERROR("PpmImage write failed during ArResolver::OpenAssetForWrite '" + filename + "'");
        return;
    }
    
    std::stringstream ss;
    ss << "P3\n";
    ss << _width << " " << _height << "\n";
    ss << _componentMaximum << "\n";
    for (size_t i = 0; i < _storage.size(); i += 3) {
        ss << _storage[i] << " " << _storage[i + 1] << " " << _storage[i + 2] << "\n";
    }
    
    std::string s = ss.str();
    
    asset->Write(s.c_str(), s.size(), 0);
    MY_VERIFY(asset->Close());
}

uint16_t PpmImage::getWidth() const { return _width; }

uint16_t PpmImage::getHeight() const { return _height; }

uint16_t PpmImage::getComponentMaximum() const { return _componentMaximum; }

size_t PpmImage::_toIndex(uint16_t x, uint16_t y, int channel) const {
    if (x >= _width || y >= _height || channel >= 3 || channel < 0) {
        MY_CODING_ERROR("Invalid coordinates or channel for PpmImage");
        return -1;
    }
    
    return y * 3 * _width + 3 * x + channel;
}

PpmImage::PpmImage() {}
